# SECTION 1: Numerical Data types and Arithmetic Operators

In [ ]:
fuel_efficiency = 11.5  # kms/liter
total_trip_distance = 475  # kms
fuel_price = 1.65       # $/liter
tank_capacity = 50      # liters


Question A:

In [ ]:
total_fuel_required = total_trip_distance / fuel_efficiency
print(total_fuel_required)

41.30434782608695


Explanation: to find out the total fuel required for the entire journey we divided total trip distance by the efficiency of the vehicle , that is 475kms by 11.5 kms/liter equals to 41.3 liters.

Question B:

In [ ]:
full_tanks = total_fuel_required/ tank_capacity
print(full_tanks)

0.826086956521739


Explanation: To figure out the number of tanks that will be required to be full for the entire trip, we divided the total fuel by the entire tank capacity in liters that results in 0.82 which indicates to 0 full tanks approximately.

  Question C:

In [ ]:
total_fuel_cost= total_fuel_required * fuel_price
print( total_fuel_cost)

68.15217391304347


Explanation: to find out the total cost associated with the fuel for the trip, we multiply the total_ fuel required by the cost of fuel per liter. 41.3kms into $ 1.65/ liter.

Question D:

In [ ]:
total_fuel_remaining= tank_capacity-total_fuel_required
print(round(total_fuel_remaining,2))

8.7


To find how much fuel was left after the end of the entire trip, we deduct the fuel required from the entire fuel tank capacity of the vehicle ,             
  i.e 50-41.3 =8.7

# **SECTION 2: Comparison and Logical Operators:**

In [ ]:
age = 30
is_weekday = True
is_family_day = False
group_size = 3


Question A:

In [ ]:
discount = ((age <= 12 and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4))
print(discount)

False


Explanation: for the above scenario-
 Age is above 12
 Dosent qualify as Senior
 It is not a family day and the group size is less than 4.
 So none of the conditions match to secure a discount.



Question B:

In [ ]:
age = 10
is_weekday = True
is_family_day = False
group_size = 2


In [ ]:
discount = ((age <= 12 and is_weekday) or (age >= 65) or (is_family_day and group_size >= 4))
print(discount)

True


 QUESTION C: Explanation: the above scenario satsifies the first condition the age of the person is less than equal to 12 and also it is a weekday.

# SECTION 3: Strings, Indexing and Slicing

Question A:

In [ ]:
secret_message = "PyThoN_Is_FuN_AnD_PoWeRfUl!"


In [ ]:
print(secret_message[5])


N


Explanation: The index ranking starts from 0,1,2 etc. therefore choosing the indes as 5 will return the 6th character of the text that is 'N'.

Question B:

In [ ]:
substring = secret_message[7:13]
print(substring)



Is_FuN


Explanation: We can use the slicing function for extracting a part of the text. So we type in 7-13(index ) 7th index inclusive and 13th index exclusive.

Question C:

In [ ]:
print(secret_message[-9:-1])

PoWeRfUl


Explanation: for using the negative index function and extracting a part of the message , we hve to define the index number as [-9 to -1] as the negative indexing starts from -1.

Question D:

In [ ]:
print(secret_message[::4])

PosNDWU


Explanation: [::4] indicates that it prints every 4th character simultaneously after each turn.

Question E:

In [ ]:
print= (secret_message[50])

NameError: name 'secret_message' is not defined

Explanation: it pops out an error becuase the string(secret message)does not have 50 characters.

# SECTION 4: Boleean Values and Expression

In [ ]:
x = 15
y = 25
z = 15
text = "adventure"


Question A:

In [ ]:
print((x == z) and (y > x))


True


Explanation: The value of x(15) is equal to the value of z(15) and the value of y(25) is greater than x. Therefore the final output is True.

  Question B:
  

In [ ]:
print((x != z) or (y < x) or (len(text) < 5))


False


Explanation: The final output is false because the value of x(15) is equal to y(25), the value of y is greater than x and the length of the string is 9 and not 5.

Question C:

In [ ]:
print(not ((x > y) and (text == "Adventure")))


True


Explanation: The value of x(15) is not greater than the value of y(250 and the text does not match the given string "adventure".

# SECTION 5: STRING PROPERTIES AND METHODS:

Question A:

In [ ]:
username = "MyUser123"
username[0] = 'm'


TypeError: 'str' object does not support item assignment

Explanation: once the string is created, the characters of the string cannot be modified using indexing function.

Question B:

In [1]:
quote = " practice makes perfect "




In [2]:
quote = quote.strip()
quote = quote.upper()
quote = quote.replace("PERFECT","PROGRESS")
print(quote)

PRACTICE MAKES PROGRESS


Explanation: By using the strip function, we removed the spaces at the beginning and end of the string.
By using the upper function we changed the character case from lower to upper and replace function changed the word from perfect to progress.

Question C:

In [3]:
original = " practice makes perfect "
formatted = quote

In [4]:
print("Original length of the string:", len(original))
print("Formatted length of the string:", len(formatted))

Original length of the string: 24
Formatted length of the string: 23


Explanation: The length of the original string was 24, After formatting the quote it is 23 as the spaces from either ends are removed.